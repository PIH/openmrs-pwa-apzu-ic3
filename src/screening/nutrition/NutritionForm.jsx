import React from 'react';
import { connect } from 'react-redux';
import { formValueSelector } from 'redux-form';
import { Obs, formValidations } from '@openmrs/react-components';
import { Alert, Grid, Row, FormGroup, ControlLabel, Label, Col } from 'react-bootstrap';
import { ENCOUNTER_TYPES, CONCEPTS, MALNUTRITION_LEVEL, FORM_ANSWERS } from "../../constants";
import utils from "../../utils";
import ScreeningForm from "../ScreeningForm";
import { colHeight, leftTextAlign } from "../../pwaStyles";

const minValue25 = formValidations.minValue(25);
const maxValue140 = formValidations.maxValue(140);
const minValue2 = formValidations.minValue(2);
const maxValue100 = formValidations.maxValue(100);

class NutritionForm extends React.Component {

  render() {
    const formContent = (
      <Grid>
        <Row>
          <FormGroup controlId="formWeight">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              Weight
            </Col>
            <Col sm={2}>
              <Obs
                concept={CONCEPTS.Weight.uuid}
                path="weight"
                placeholder="weight in kg"
                validate={this.props.patient.age > 18 ? [minValue25, maxValue140] :  [minValue2, maxValue100]}
              />
            </Col>
            <Col
              componentClass={ControlLabel}
              sm={2}
              style={leftTextAlign}
            >
              Kg
            </Col>
          </FormGroup>
        </Row>
        <Row>
          <FormGroup controlId="formHeight">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              Height
            </Col>
            <Col sm={2}>
              <Obs
                concept={CONCEPTS.Height.uuid}
                path="height"
                placeholder="height in cm"
              />
            </Col>
            <Col
              componentClass={ControlLabel}
              sm={2}
              style={leftTextAlign}
            >
              cm
            </Col>
          </FormGroup>
        </Row>

        {( (this.props.patient.age >= 18) && (this.props.patient.age < 50) && (this.props.patient.gender === 'F') ) &&
        <Row>
          <FormGroup controlId="formPregnant">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
            Pregnant
            </Col>
            <Col sm={4}>
              <Obs
                concept={CONCEPTS.Pregnant.uuid}
                conceptAnswers={FORM_ANSWERS.trueFalse}
                path="pregnant"
              />
            </Col>
          </FormGroup>
        </Row>
        }

        { (( this.props.patient.age >= 18 && this.props.patient.gender === 'M')
          || (this.props.patient.gender === 'F' && this.props.patient.age >= 18 && (this.props.pregnant !== null) && (this.props.pregnant === CONCEPTS.False.uuid))
          || (this.props.patient.gender === 'F' && this.props.patient.age > 50))&&
          <Row>
            <FormGroup controlId="formBMI">
              <Col
                componentClass={ControlLabel}
                sm={2}
              >
              BMI
              </Col>
              { (typeof this.props.weight !== 'undefined') && (typeof this.props.height !== 'undefined') &&
              <Col sm={2}>
                <h3><Label bsStyle={this.props.bmiStyle.alert}>{ this.props.bmi }</Label>
                </h3>
              </Col>
              }
            </FormGroup>
          </Row>
        }

        { ((this.props.patient.age < 18) || ( (this.props.patient.age >= 18) && (this.props.patient.gender === 'F') && (this.props.pregnant !== null) && (this.props.pregnant === CONCEPTS.True.uuid))) &&
        <Row>
          <FormGroup controlId="formMuac">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              MUAC
            </Col>
            <Col sm={2}>
              <Obs
                concept={CONCEPTS.MUAC.uuid}
                path="muac"
                placeholder="muac in cm"
              />
            </Col>
            <Col
              componentClass={ControlLabel}
              sm={2}
              style={leftTextAlign}
            >
              cm
            </Col>
          </FormGroup>
        </Row>
        }
        {(this.props.malnutrition !== null)
        && ((this.props.malnutrition === MALNUTRITION_LEVEL.severe) || (this.props.malnutrition === MALNUTRITION_LEVEL.moderate)) &&
        <Row>
          <FormGroup controlId="formMalnutrition">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              Malnutrition
            </Col>
            <Col sm={2}>
              <Alert bsStyle={this.props.malnutrition.alert}>
                { this.props.malnutrition.message }
              </Alert>
            </Col>

          </FormGroup>
        </Row>
        }
        <Row>
          <Col
            md={20}
            sm={20}
            style={colHeight}
          >
            <span><h1>{ '' }</h1></span>
          </Col>
        </Row>
      </Grid>
    );

    return (
      <ScreeningForm
        encounterType={ENCOUNTER_TYPES.NutritionEncounterType}
        formContent={formContent}
        formId="nutrition-form"
        formInstanceId="nutrition-form"
        title="Nutrition"
      />
    );
  }

}

const selector = formValueSelector('nutrition-form');

export default connect(state => {
  const patient = state.openmrs.selectedPatient ? state.openmrs.patients[state.openmrs.selectedPatient] : null;
  const weight = selector(state, 'obs|path=weight|concept=' + CONCEPTS.Weight.uuid);
  const height = selector(state, 'obs|path=height|concept=' + CONCEPTS.Height.uuid);
  const muac = selector(state, 'obs|path=muac|concept=' + CONCEPTS.MUAC.uuid);
  const pregnant = selector(state, 'obs|path=pregnant|concept=' + CONCEPTS.Pregnant.uuid);
  let bmi = null;
  let bmiStyle = MALNUTRITION_LEVEL.none;
  if ( typeof weight !== 'undefined' && typeof height !== 'undefined') {
    bmi = utils.calculateBMI(weight, height);
    bmiStyle = utils.calculateBMIAlert(bmi);
  }
  const malnutrition = utils.calculateMalnutritionLevel(bmi, muac, patient ? patient.age : null, pregnant);
  return {
    weight,
    height,
    pregnant,
    bmi,
    bmiStyle,
    malnutrition,
    patient,
  };
})(NutritionForm);

