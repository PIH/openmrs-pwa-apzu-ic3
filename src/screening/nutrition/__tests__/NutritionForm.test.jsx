import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import MockRouter from 'react-mock-router';
import NutritionForm from '../NutritionForm';

let props, store;
let mountedComponent;

const mockStore = configureMockStore();

const nutritionForm = () => {
  if (!mountedComponent) {
    mountedComponent = mount(
      <Provider store={store}>
        <MockRouter>
          <NutritionForm {...props} />
        </MockRouter>
      </Provider>);
  }
  return mountedComponent;
};

describe('Component: NutritionForm', () => {
  beforeEach(() => {
    props = {};
    store = mockStore(
      {
        selectedPatient: {
          uuid: "abc",
          visit: {
            uuid: "def"
          }
        }
      });
    mountedComponent = undefined;
  });

  it('renders properly', () => {
    expect(toJson(nutritionForm())).toMatchSnapshot();
  });

});
