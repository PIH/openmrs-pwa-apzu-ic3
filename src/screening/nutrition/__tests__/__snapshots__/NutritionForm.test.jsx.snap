// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: NutritionForm renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MockRouter
    createHref={[Function]}
    location=""
    params={Object {}}
    path=""
    push={[Function]}
    replace={[Function]}
    url=""
  >
    <Connect(NutritionForm)>
      <NutritionForm
        bmi={null}
        bmiStyle={
          Object {
            "alert": "default",
            "message": "None",
          }
        }
        dispatch={[Function]}
        malnutrition={null}
        patient={
          Object {
            "uuid": "abc",
            "visit": Object {
              "uuid": "def",
            },
          }
        }
      >
        <Connect(ScreeningForm)
          encounterType={
            Object {
              "uuid": "6265F6BC-EBC0-4181-91F3-28B70BBFDB61",
            }
          }
          formContent={
            <Grid
              bsClass="container"
              componentClass="div"
              fluid={false}
            >
              <Row
                bsClass="row"
                componentClass="div"
              >
                <FormGroup
                  bsClass="form-group"
                  controlId="formWeight"
                >
                  <Col
                    bsClass="col"
                    componentClass={[Function]}
                    sm={2}
                  >
                    Weight
                  </Col>
                  <Col
                    bsClass="col"
                    componentClass="div"
                    sm={2}
                  >
                    <WithObsGroupContext(WithFormContext(Connect(Obs)))
                      concept="6569c44a-977f-11e1-8993-905e29aff6c1"
                      path="weight"
                      placeholder="weight in kg"
                      validate={
                        Array [
                          [Function],
                          [Function],
                        ]
                      }
                    />
                  </Col>
                  <Col
                    bsClass="col"
                    componentClass={[Function]}
                    sm={2}
                    style={
                      Object {
                        "textAlign": "left",
                      }
                    }
                  >
                    Kg
                  </Col>
                </FormGroup>
              </Row>
              <Row
                bsClass="row"
                componentClass="div"
              >
                <FormGroup
                  bsClass="form-group"
                  controlId="formHeight"
                >
                  <Col
                    bsClass="col"
                    componentClass={[Function]}
                    sm={2}
                  >
                    Height
                  </Col>
                  <Col
                    bsClass="col"
                    componentClass="div"
                    sm={2}
                  >
                    <WithObsGroupContext(WithFormContext(Connect(Obs)))
                      concept="6569c562-977f-11e1-8993-905e29aff6c1"
                      path="height"
                      placeholder="height in cm"
                    />
                  </Col>
                  <Col
                    bsClass="col"
                    componentClass={[Function]}
                    sm={2}
                    style={
                      Object {
                        "textAlign": "left",
                      }
                    }
                  >
                    cm
                  </Col>
                </FormGroup>
              </Row>
              <Row
                bsClass="row"
                componentClass="div"
              >
                <Col
                  bsClass="col"
                  componentClass="div"
                  md={20}
                  sm={20}
                  style={
                    Object {
                      "height": "40px",
                    }
                  }
                >
                  <span>
                    <h1 />
                  </span>
                </Col>
              </Row>
            </Grid>
          }
          formId="nutrition-form"
          formInstanceId="nutrition-form"
          title="Nutrition"
        >
          <ScreeningForm
            dispatch={[Function]}
            encounterType={
              Object {
                "uuid": "6265F6BC-EBC0-4181-91F3-28B70BBFDB61",
              }
            }
            formContent={
              <Grid
                bsClass="container"
                componentClass="div"
                fluid={false}
              >
                <Row
                  bsClass="row"
                  componentClass="div"
                >
                  <FormGroup
                    bsClass="form-group"
                    controlId="formWeight"
                  >
                    <Col
                      bsClass="col"
                      componentClass={[Function]}
                      sm={2}
                    >
                      Weight
                    </Col>
                    <Col
                      bsClass="col"
                      componentClass="div"
                      sm={2}
                    >
                      <WithObsGroupContext(WithFormContext(Connect(Obs)))
                        concept="6569c44a-977f-11e1-8993-905e29aff6c1"
                        path="weight"
                        placeholder="weight in kg"
                        validate={
                          Array [
                            [Function],
                            [Function],
                          ]
                        }
                      />
                    </Col>
                    <Col
                      bsClass="col"
                      componentClass={[Function]}
                      sm={2}
                      style={
                        Object {
                          "textAlign": "left",
                        }
                      }
                    >
                      Kg
                    </Col>
                  </FormGroup>
                </Row>
                <Row
                  bsClass="row"
                  componentClass="div"
                >
                  <FormGroup
                    bsClass="form-group"
                    controlId="formHeight"
                  >
                    <Col
                      bsClass="col"
                      componentClass={[Function]}
                      sm={2}
                    >
                      Height
                    </Col>
                    <Col
                      bsClass="col"
                      componentClass="div"
                      sm={2}
                    >
                      <WithObsGroupContext(WithFormContext(Connect(Obs)))
                        concept="6569c562-977f-11e1-8993-905e29aff6c1"
                        path="height"
                        placeholder="height in cm"
                      />
                    </Col>
                    <Col
                      bsClass="col"
                      componentClass={[Function]}
                      sm={2}
                      style={
                        Object {
                          "textAlign": "left",
                        }
                      }
                    >
                      cm
                    </Col>
                  </FormGroup>
                </Row>
                <Row
                  bsClass="row"
                  componentClass="div"
                >
                  <Col
                    bsClass="col"
                    componentClass="div"
                    md={20}
                    sm={20}
                    style={
                      Object {
                        "height": "40px",
                      }
                    }
                  >
                    <span>
                      <h1 />
                    </span>
                  </Col>
                </Row>
              </Grid>
            }
            formId="nutrition-form"
            formInstanceId="nutrition-form"
            patient={
              Object {
                "uuid": "abc",
                "visit": Object {
                  "uuid": "def",
                },
              }
            }
            title="Nutrition"
          >
            <Connect(EncounterFormPage)
              encounterRole={
                Object {
                  "uuid": "a0b03050-c99b-11e0-9572-0800200c9a66",
                }
              }
              encounterType={
                Object {
                  "uuid": "6265F6BC-EBC0-4181-91F3-28B70BBFDB61",
                }
              }
              formContent={
                <Grid
                  bsClass="container"
                  componentClass="div"
                  fluid={false}
                >
                  <Row
                    bsClass="row"
                    componentClass="div"
                  >
                    <FormGroup
                      bsClass="form-group"
                      controlId="formWeight"
                    >
                      <Col
                        bsClass="col"
                        componentClass={[Function]}
                        sm={2}
                      >
                        Weight
                      </Col>
                      <Col
                        bsClass="col"
                        componentClass="div"
                        sm={2}
                      >
                        <WithObsGroupContext(WithFormContext(Connect(Obs)))
                          concept="6569c44a-977f-11e1-8993-905e29aff6c1"
                          path="weight"
                          placeholder="weight in kg"
                          validate={
                            Array [
                              [Function],
                              [Function],
                            ]
                          }
                        />
                      </Col>
                      <Col
                        bsClass="col"
                        componentClass={[Function]}
                        sm={2}
                        style={
                          Object {
                            "textAlign": "left",
                          }
                        }
                      >
                        Kg
                      </Col>
                    </FormGroup>
                  </Row>
                  <Row
                    bsClass="row"
                    componentClass="div"
                  >
                    <FormGroup
                      bsClass="form-group"
                      controlId="formHeight"
                    >
                      <Col
                        bsClass="col"
                        componentClass={[Function]}
                        sm={2}
                      >
                        Height
                      </Col>
                      <Col
                        bsClass="col"
                        componentClass="div"
                        sm={2}
                      >
                        <WithObsGroupContext(WithFormContext(Connect(Obs)))
                          concept="6569c562-977f-11e1-8993-905e29aff6c1"
                          path="height"
                          placeholder="height in cm"
                        />
                      </Col>
                      <Col
                        bsClass="col"
                        componentClass={[Function]}
                        sm={2}
                        style={
                          Object {
                            "textAlign": "left",
                          }
                        }
                      >
                        cm
                      </Col>
                    </FormGroup>
                  </Row>
                  <Row
                    bsClass="row"
                    componentClass="div"
                  >
                    <Col
                      bsClass="col"
                      componentClass="div"
                      md={20}
                      sm={20}
                      style={
                        Object {
                          "height": "40px",
                        }
                      }
                    >
                      <span>
                        <h1 />
                      </span>
                    </Col>
                  </Row>
                </Grid>
              }
              formId="nutrition-form"
              formInstanceId="nutrition-form"
              formSubmittedActionCreators={
                Array [
                  [Function],
                ]
              }
              title="Nutrition"
            >
              <EncounterFormPage
                dispatch={[Function]}
                encounterRole={
                  Object {
                    "uuid": "a0b03050-c99b-11e0-9572-0800200c9a66",
                  }
                }
                encounterType={
                  Object {
                    "uuid": "6265F6BC-EBC0-4181-91F3-28B70BBFDB61",
                  }
                }
                formContent={
                  <Grid
                    bsClass="container"
                    componentClass="div"
                    fluid={false}
                  >
                    <Row
                      bsClass="row"
                      componentClass="div"
                    >
                      <FormGroup
                        bsClass="form-group"
                        controlId="formWeight"
                      >
                        <Col
                          bsClass="col"
                          componentClass={[Function]}
                          sm={2}
                        >
                          Weight
                        </Col>
                        <Col
                          bsClass="col"
                          componentClass="div"
                          sm={2}
                        >
                          <WithObsGroupContext(WithFormContext(Connect(Obs)))
                            concept="6569c44a-977f-11e1-8993-905e29aff6c1"
                            path="weight"
                            placeholder="weight in kg"
                            validate={
                              Array [
                                [Function],
                                [Function],
                              ]
                            }
                          />
                        </Col>
                        <Col
                          bsClass="col"
                          componentClass={[Function]}
                          sm={2}
                          style={
                            Object {
                              "textAlign": "left",
                            }
                          }
                        >
                          Kg
                        </Col>
                      </FormGroup>
                    </Row>
                    <Row
                      bsClass="row"
                      componentClass="div"
                    >
                      <FormGroup
                        bsClass="form-group"
                        controlId="formHeight"
                      >
                        <Col
                          bsClass="col"
                          componentClass={[Function]}
                          sm={2}
                        >
                          Height
                        </Col>
                        <Col
                          bsClass="col"
                          componentClass="div"
                          sm={2}
                        >
                          <WithObsGroupContext(WithFormContext(Connect(Obs)))
                            concept="6569c562-977f-11e1-8993-905e29aff6c1"
                            path="height"
                            placeholder="height in cm"
                          />
                        </Col>
                        <Col
                          bsClass="col"
                          componentClass={[Function]}
                          sm={2}
                          style={
                            Object {
                              "textAlign": "left",
                            }
                          }
                        >
                          cm
                        </Col>
                      </FormGroup>
                    </Row>
                    <Row
                      bsClass="row"
                      componentClass="div"
                    >
                      <Col
                        bsClass="col"
                        componentClass="div"
                        md={20}
                        sm={20}
                        style={
                          Object {
                            "height": "40px",
                          }
                        }
                      >
                        <span>
                          <h1 />
                        </span>
                      </Col>
                    </Row>
                  </Grid>
                }
                formId="nutrition-form"
                formInstanceId="nutrition-form"
                formSubmittedActionCreators={
                  Array [
                    [Function],
                  ]
                }
                patient={
                  Object {
                    "uuid": "abc",
                    "visit": Object {
                      "uuid": "def",
                    },
                  }
                }
                title="Nutrition"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                      "paddingRight": "0px",
                    }
                  }
                >
                  <Grid
                    bsClass="container"
                    componentClass="div"
                    fluid={false}
                    style={
                      Object {
                        "paddingLeft": "0px",
                        "paddingRight": "0px",
                      }
                    }
                  >
                    <div
                      className="container"
                      style={
                        Object {
                          "paddingLeft": "0px",
                          "paddingRight": "0px",
                        }
                      }
                    >
                      <Row
                        bsClass="row"
                        componentClass="div"
                        style={
                          Object {
                            "backgroundColor": "#ffa500b3",
                          }
                        }
                      >
                        <div
                          className="row"
                          style={
                            Object {
                              "backgroundColor": "#ffa500b3",
                            }
                          }
                        >
                          <Col
                            bsClass="col"
                            componentClass="div"
                            md={20}
                            sm={20}
                            style={
                              Object {
                                "paddingLeft": "20px",
                              }
                            }
                          >
                            <div
                              className="col-md-20 col-sm-20"
                              style={
                                Object {
                                  "paddingLeft": "20px",
                                }
                              }
                            >
                              <span>
                                <h1>
                                  Nutrition
                                </h1>
                              </span>
                            </div>
                          </Col>
                        </div>
                      </Row>
                      <Row
                        bsClass="row"
                        componentClass="div"
                      >
                        <div
                          className="row"
                        >
                          <Col
                            bsClass="col"
                            componentClass="div"
                            md={20}
                            sm={20}
                            style={
                              Object {
                                "height": "40px",
                              }
                            }
                          >
                            <div
                              className="col-md-20 col-sm-20"
                              style={
                                Object {
                                  "height": "40px",
                                }
                              }
                            >
                              <span>
                                <h1 />
                              </span>
                            </div>
                          </Col>
                        </div>
                      </Row>
                    </div>
                  </Grid>
                  <Loader>
                    <div
                      className="custom-loader"
                    >
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                      <div />
                    </div>
                  </Loader>
                </div>
              </EncounterFormPage>
            </Connect(EncounterFormPage)>
          </ScreeningForm>
        </Connect(ScreeningForm)>
      </NutritionForm>
    </Connect(NutritionForm)>
  </MockRouter>
</Provider>
`;
